server/src/test/java/org/elasticsearch/cluster/metadata/MetadataCreateIndexServiceTests.java

```diff
@@ -691,6 +694,178 @@ public void testAggregateSettingsAppliesSettingsFromTemplatesAndRequest() {
         assertThat(aggregatedIndexSettings.get("request_setting"), equalTo("value2"));
     }
 
+    public void testAggregateSettingsProviderOverrulesSettingsFromRequest() {
```

Comment:
These unit tests look good. But I think we should also have a test similar to `IndexSettingProviderTests`. Then we can really test in the context of index creation, template creation and simulate index template api.

Reply:
Hm the provider doesn't have logic related to overruling, so these unit tests belong here imho. We can duplicate them, but they don't seem to add testing coverage.

Reply:
So this would be more of an integration test. Which allows the testing of an overruling index settings provider in the context of index creation, put template action and simulate index template actions.

Reply:
I think it's cleaner to keep these tests separately, in the corresponding unit tests. It's confusing to have tests for these actions here, since there's no associated logic.

Reply:
Adding a test like `IndexSettingProviderTests` where we test that one or more index settings get overruled in context index creation, template validation and simulate index template action does increase test coverage.

For example the use case where an index template gets added that sets a setting that gets overruled would now only happen upon index creation and not during template validation in this change. Which could cause potential validation errors to happen not during template creation / update, but after new indices get created. A test that validates that settings also get overruled during template validation can avoid such issues.

When adding or changing plugin infrastructure, I think integration level testing is something that improves code quality.

Reply:
These cases can easily be covered in `MetadataIndexTemplateServiceTests` and `TransportSimulateIndexTemplateActionTests` where the corresponding logic lies? 

But - see my other comment. Do we really want validation for templates? Templates can be created before licence changes, so validation should not apply to them imho - setting validation and overruling should only happen at the time of index creation.

Reply:
> These cases can easily be covered in MetadataIndexTemplateServiceTests and TransportSimulateIndexTemplateActionTests where the corresponding logic lies?

These are unit tests. And we can add tests to these test suites. But I think we should also have integration tests.

> But - see my other comment. Do we really want validation for templates? Templates can be created before licence changes, so validation should not apply to them imho - setting validation and overruling should only happen at the time of index creation.

Yes, I think we want that. Although we make this change to `IndexSettingsProvider` for the synthetic source fallback logic. We need to test the overruling logic in general. It may get used in other contexts down the road.

I think having potential validation errors because of index settings providers at index creation time would be a bad user experience. It is better to have validation errors at template creation / update.

Reply:
Wrt to the test, I agree it's a good idea to add it here - done.